### PR TITLE
[lldb] Create dependent modules in parallel

### PR DESCRIPTION
**Pull request by JDevlieghere**
Create dependent modules in parallel in Target::SetExecutableModule. This change was inspired by #110646 which takes the same approach when attaching. Jason suggested we could use the same approach when you create a target in LLDB.

I used Slack for benchmarking, which loads 902 images.

```
Benchmark 1: ./bin/lldb /Applications/Slack.app/Contents/MacOS/Slack
  Time (mean ± σ):      1.225 s ±  0.003 s    [User: 3.977 s, System: 1.521 s]
  Range (min … max):    1.220 s …  1.229 s    10 runs

Benchmark 2: ./bin/lldb /Applications/Slack.app/Contents/MacOS/Slack
  Time (mean ± σ):      3.253 s ±  0.037 s    [User: 3.013 s, System: 0.248 s]
  Range (min … max):    3.211 s …  3.310 s    10 runs
```

We see about a 2x speedup, which matches what Jason saw for the attach scenario. I also ran this under TSan to confirm this doesn't introduce any races or deadlocks.

### CODE REVIEW

**Comment by JDevlieghere** [comment]
CC @DmT021

File: lldb/source/Target/Target.cpp [hunk 1]
```diff
@@ -1608,9 +1612,28 @@ void Target::SetExecutableModule(ModuleSP &executable_sp,
         if (image_module_sp) {
           added_modules.AppendIfNeeded(image_module_sp, false);
           ObjectFile *objfile = image_module_sp->GetObjectFile();
-          if (objfile)
+          if (objfile) {
+            std::lock_guard<std::mutex> guard(dependent_files_mutex);
             objfile->GetDependentModules(dependent_files);
```
Comment (DmT021) [inline lldb/source/Target/Target.cpp:1617]: I wonder if this operation is heavy in any of the `ObjectFile` implementations. If it is we may want to lock the mutex only when an actual append to the dependent_files happens.

**Review by jasonmolenda — APPROVED** [review]
It took me a minute to understand how you were handling the locking of the dependent file list while iterating across it, but now I see it.  You grab the lock at the beginning of the outer loop, then enqueue all of the not-yet-processed dependent binaries into asynchronous threads.  They are all creating the Modules and then blocking to acquire the lock, to add their entries to the dependent files list.  The inner loop finishes distributing things to the thread pool, releases its lock, and waits for all the just-started async jobs to complete.  Then we are back in the outer loop which reads the size of the dependent file list.

**Comment by jasonmolenda** [comment]
Looks good to me, thanks for addressing this other bottleneck for launch startup.

File: lldb/source/Target/Target.cpp (hunk 1 above)
Comment (jasonmolenda) [reply lldb/source/Target/Target.cpp:1617]: good question.  for mach-o this is a simple iteration over the load commands, but we might want to to pass a temporaryfile list object to this method, and then acquire the lock and append its entries to `dependency_files`, if we didn't want to assume that.

**Comment by JDevlieghere** [comment]
The test failures were both caused by the locking in `PathMappingList`, which I'm addressing in #114576.

File: lldb/source/Target/Target.cpp (hunk 1 above)
Comment (JDevlieghere) [reply lldb/source/Target/Target.cpp]: Yeah fair enough. Even the Mach-O one is doing some filesystem operations that really shouldn't be happening while holding the lock. I don't think we need to make every FileSpecList thread safe, and passing an optional `mutex*` was messy, so I went with the local copy as Jason suggested.

File: lldb/source/Target/Target.cpp [hunk 2]
```diff
@@ -1608,9 +1612,48 @@ void Target::SetExecutableModule(ModuleSP &executable_sp,
...
+      llvm::ThreadPoolTaskGroup task_group(Debugger::GetThreadPool());
+      for (uint32_t i = 0; i < dependent_files.GetSize(); i++) {
+        // Process all currently known dependencies in parallel in the innermost
+        // loop. This may create newly discovered dependencies to be appended to
+        // dependent_files. We'll deal with these files during the next
+        // iteration of the outermost loop.
+        {
+          std::lock_guard<std::mutex> guard(dependent_files_mutex);
+          for (; i < dependent_files.GetSize(); i++)
+            task_group.async(GetDependentModules,
```
Comment (adrian-prantl) [inline lldb/source/Target/Target.cpp:1653]: async inside a lock_guard looks dangerous. I assume you're sure this won't recurse into this function? :-)

**Review by jasonmolenda — APPROVED** [review]
LGTM.

File: lldb/source/Target/Target.cpp (hunk 1 above)
Comment (DmT021) [reply lldb/source/Target/Target.cpp]: Just in case, did you check for a regression in performance after the latest patch?

Comment (JDevlieghere) [reply lldb/source/Target/Target.cpp]: Yeah I did. I was hoping for a slight speedup, but the difference was in the noise.

File: lldb/source/Target/Target.cpp (hunk 2 above)
Comment (JDevlieghere) [reply lldb/source/Target/Target.cpp:1653]: It doesn't, but if that becomes a problem we can do the same trick that Jason suggested and iterate over a copy of the list here. For now that seems unnecessary but hopefully someone will see this comment if that ever changes :-)

**Pull request merged**